packages/flutter/lib/src/cupertino/sliding_segmented_control.dart

```diff
@@ -357,111 +480,236 @@ class _SegmentedControlState<T> extends State<CupertinoSlidingSegmentedControl<T
     super.dispose();
   }
 
-  // Play highlight animation for the child located at _highlightControllers[at].
-  void _animateHighlightController({ T? at, required bool forward }) {
-    if (at == null)
-      return;
-    final AnimationController? controller = _highlightControllers[at];
-    assert(!forward || controller != null);
-    controller?.animateTo(forward ? 1 : 0, duration: _kHighlightAnimationDuration, curve: Curves.ease);
+  // Whether the current drag gesture started on a selected segment. When this
+  // flag is false, the `onUpdate` method does not update `highlighted`.
+  // Otherwise the thumb can be dragged around in an ongoing drag gesture.
+  bool? _startedOnSelectedSegment;
+
+  // Whether an ongoing horizontal drag gesture that started on the thumb is
+  // present. When true, defer/ignore changes to the `highlighted` variable
+  // from other sources (except for semantics) until the gesture ends, preventing
+  // them from interfering with the active drag gesture.
+  bool get isThumbDragging => _startedOnSelectedSegment ?? false;
+
+  T segmentForXPosition(double dx) {
+    final RenderBox renderBox = context.findRenderObject()! as RenderBox;
+    final int numOfChildren = widget.children.length;
+    assert(renderBox.hasSize);
+    assert(numOfChildren >= 2);
+    int index = (dx ~/ (renderBox.size.width / numOfChildren)).clamp(0, numOfChildren - 1);
+
+    switch (Directionality.of(context)) {
+      case TextDirection.ltr:
+        break;
+      case TextDirection.rtl:
+        index = numOfChildren - 1 - index;
+        break;
+    }
+
+    assert(index < widget.children.length);
+    for (final T key in widget.children.keys) {
+      if (index == 0)
+        return key;
+      index -= 1;
+    }
```

Comment:
Couldn't you do `return widget.children.keys.elementAt(index);` and remove the assert and return below this?  Or maybe I'm missing something...